Fix emptyDir confuse gitRepo

resouer opened the pull request:
Fixes: #18167

We need to set name of wrapper volume, so plugins will not "share" the same empty_dir they rely on.

ncdc commented:
cc @pmorie

pmorie commented on pkg/volume/empty_dir/empty_dir.go:
```diff
@@ -158,14 +158,15 @@ func (ed *emptyDir) SetUpAt(dir string) error {
-	// medium is memory, and a mountpoint is present, then the volume is
-	// ready.
+	// Check the plugin readiness file is present for this volume.
+	// If the storage medium is the default, and the expexted vol path already existed,
```
nit: expected

resouer commented:
e2e should not die ... just a minor fix

anyway, I prefer to see if this modify has potential bad affects?

resouer committed:
Fix volume wrapper doesn't have name

Integration test

thockin commented on pkg/volume/git_repo/git_repo.go:
```diff
@@ -126,10 +126,8 @@ func (b *gitRepoVolumeBuilder) SetUp() error {
-	Volume: &api.Volume{VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{}}},
-}
+// This is the name for the volume that this plugin wraps, will be generated with random value in SetUpAt(..).
+var wrapperVolumeName string
```
This can't be a global.  Every call to SetUpAt() will overwrite it, which means TearDown() will fail.  It has to be attached to each volume.  But then random doesn't work because Builder/Cleaner have different amounts of information.

I think it has to be determinable from the volName.  I left us an affordance here - the API enforces that volume names are DNS_LABEL formatted.  That gives us (I hope) freedome to use things like _ in the implementation (which we do in Docker container names, for example).

What if we simply named the wrapee "wrapped_" + volName ?

thockin replied:
We could even embed that into NewWrapperBuilder & Cleaner() - just pass in volName and those can figure out the wrappee's name.

resouer replied:
I agree with wrapped_+volName, if we can not use global

resouer replied:
Sure, will add a integration test in next commit

resouer replied:
Note for me, downwardAPI also exposed to this issue, thanks Paul.

pmorie replied:
I guess I'm wondering, why not just use the volume name?  Also, emptyDir should add a validation that the volume name is set.

thockin replied:
I didn't want to use volume name on the off chance that there was ever a
mode of one volume that wraps another mode of the same volume, or that
anyone might make a list of all volume names and get a conflict.  Using an
"illegal" value means we control the prefix.

On Fri, Dec 11, 2015 at 10:25 AM, Paul Morie <notifications@github.com>
wrote:

> In pkg/volume/git_repo/git_repo.go
> <https://github.com/kubernetes/kubernetes/pull/18445#discussion_r47386661>
> :
>
> > @@ -126,10 +126,8 @@ func (b *gitRepoVolumeBuilder) SetUp() error {
> >  	return b.SetUpAt(b.GetPath())
> >  }
> >
> > -// This is the spec for the volume that this plugin wraps.
> > -var wrappedVolumeSpec = &volume.Spec{
> > -	Volume: &api.Volume{VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{}}},
> > -}
> > +// This is the name for the volume that this plugin wraps, will be generated with random value in SetUpAt(..).
> > +var wrapperVolumeName string
>
> I guess I'm wondering, why not just use the volume name? Also, emptyDir
> should add a validation that the volume name is set.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18445/files#r47386661>.
>

thockin commented:
Also, this needs at least an integration test

resouer commented on pkg/volume/volume.go:
```diff
@@ -98,3 +104,7 @@ func RenameDirectory(oldPath, newName string) (string, error) {
 }
+
+func GetWrapperVolumeName(volName string) string {
+	return fmt.Sprintf("%v-wrapper-%v", volName, rand.String(randomLength))
```
@thockin It's not a random name, it's "volumeName-wrapper-random"

pmorie commented:
@resouer first, let me say classy digging for finding this :100:

pmorie commented:
@resouer the downward API volume plugin is also affected by this.

resouer commented:
Integration test has no root permission, will make tests to e2e

resouer committed:
Add e2e test for wrapper volumes

Use git server in e2e and refactor wrapper builder

Update e2e test to use a fake git server

resouer commented:
@k8s-bot  test this

resouer commented:
@k8s-bot test this

pmorie commented on pkg/volume/downwardapi/downwardapi_test.go:
```diff
@@ -141,6 +143,17 @@ func TestLabels(t *testing.T) {
 		t.Errorf("Failed to setup volume: %v", err)
 	}
 
+	// gitRepo volume should create it's own empty wrapper path
```
typo: its (english is weird)

pmorie replied:
also, copy-paste error re: git

pmorie commented on pkg/volume/git_repo/git_repo_test.go:
```diff
@@ -263,6 +266,19 @@ func doTestPlugin(scenario struct {
 		}
 	}
 
+	// gitRepo volume should create it's own empty wrapper path
```
its

pmorie commented on pkg/volume/secret/secret_test.go:
```diff
@@ -110,6 +110,16 @@ func TestPlugin(t *testing.T) {
 		}
 	}
 
+	// secret volume should create it's own empty wrapper path
```
its

pmorie commented on test/e2e/empty_dir_wrapper.go:
```diff
@@ -0,0 +1,122 @@
+						Name: gitVolumeName,
+						VolumeSource: api.VolumeSource{
+							GitRepo: &api.GitRepoVolumeSource{
+								Repository: "https://github.com/kubernetes/kubedash",
```
I'm a little iffy on adding a dependency on github being up for this test to pass.  I guess the alternative is running a git server in a pod and cloning from that, which is a better test scheme, but much more complex.  We do need a test for this, and integration isn't suitable because it will require the tests to have mount capability, which you don't want to have to count on for those tests.

@thockin @quinton-hoole any thoughts about this?

thockin replied:
Can't we just test secret vs downward-api - they should both follow the same pattern

pmorie replied:
Good idea

On Mon, Dec 14, 2015 at 12:58 PM, Tim Hockin <notifications@github.com>
wrote:

> In test/e2e/empty_dir_wrapper.go
> <https://github.com/kubernetes/kubernetes/pull/18445#discussion_r47532069>
> :
>
> > +			},
> > +			Spec: api.PodSpec{
> > +				Volumes: []api.Volume{
> > +					{
> > +						Name: volumeName,
> > +						VolumeSource: api.VolumeSource{
> > +							Secret: &api.SecretVolumeSource{
> > +								SecretName: name,
> > +							},
> > +						},
> > +					},
> > +					{
> > +						Name: gitVolumeName,
> > +						VolumeSource: api.VolumeSource{
> > +							GitRepo: &api.GitRepoVolumeSource{
> > +								Repository: "https://github.com/kubernetes/kubedash",
>
> Can't we just test secret vs downward-api - they should both follow the
> same pattern
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18445/files#r47532069>.
>

resouer replied:
Will fix these nits in next commit

resouer replied:
@pmorie I don't think secret vs downward-api helps, they are both `VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}}`, they will not pass `!notMnt` check. 

Set up a git server pod is a little complicated, as we are requiring some container dependency here ...but I can do that if it's the only way ... :(

thockin commented on pkg/volume/volume.go:
```diff
@@ -132,3 +133,22 @@ func RenameDirectory(oldPath, newName string) (string, error) {
+
+	var volumeSpec *Spec
+	switch pluginName {
+	case "kubernetes.io/downward-api", "kubernetes.io/secret":
```
sorry, this is problematic - these volume plugins should be completely opaque to the plugin-support logic.  The core code may never spell out or know anything about the details of any plugin.  You have to pretend that VolumeSource is an opaque blob.

The way I thought of this was that wrappers call `NewWrapperBuilder(name string, spec volume.Spec, ...)` and `NewWrapperCleaner(name string, spec volume.Spec)` and those functions are responsible for formatting the wrap-name and inserting it into the struct before passing it along.

Note that they have to take by value instead of by pointer, so they can mutate the struct (or else they have to make a copy internally).

resouer replied:
So does it do harm if we make a copy internally? I think changing type of `spec` will affect behavior of other irrelevant plugins.

resouer commented:
@pmorie Updated code and tests

pmorie commented:
@resouer rebase needed

pmorie commented on pkg/volume/downwardapi/downwardapi_test.go:
```diff
@@ -141,6 +143,17 @@ func TestLabels(t *testing.T) {
+
+	if _, err := os.Stat(podWrapperMetadataDir); err != nil {
+		if os.IsNotExist(err) {
+			t.Errorf("SetUp() failed, empty-dir wrapper path is not created: %s", podWrapperMetadataDir)
```
s/is/was

resouer commented:
@pmorie Updated, thanks for your time.

thockin commented:
Looks good to me, I give Paul final approval.

pmorie commented on test/e2e/empty_dir_wrapper.go:
```diff
@@ -0,0 +1,201 @@
+				Containers: []api.Container{
+					{
+						Name:            "git-repo",
+						Image:           "hyzual/gitolite-http",
```
@thockin Do we have any kind of policy on using third-party images in E2Es?

pmorie replied:
I'm wondering whether we should make our own image by just adding a layer on top of this one -- that way we can control version changes.

thockin replied:
We try not to rely on 3rd party images.  We can pull it, tag it, and push
to gcr
On Jan 12, 2016 8:05 AM, "Paul Morie" <notifications@github.com> wrote:

> In test/e2e/empty_dir_wrapper.go
> <https://github.com/kubernetes/kubernetes/pull/18445#discussion_r49473359>
> :
>
> > +		}
> > +
> > +		gitServerPodName := "git-server-" + string(util.NewUUID())
> > +
> > +		labels := map[string]string{"name": gitServerPodName}
> > +
> > +		gitServerPod := &api.Pod{
> > +			ObjectMeta: api.ObjectMeta{
> > +				Name:   gitServerPodName,
> > +				Labels: labels,
> > +			},
> > +			Spec: api.PodSpec{
> > +				Containers: []api.Container{
> > +					{
> > +						Name:            "git-repo",
> > +						Image:           "hyzual/gitolite-http",
>
> @thockin <https://github.com/thockin> Do we have any kind of policy on
> using third-party images in E2Es?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18445/files#r49473359>.
>

resouer replied:
I don't think I have permission to push that to gcr, please help me to tag it if possible ...

thockin replied:
I've pushed it to `gcr.io/google_containers/gitolite-http:604c51b84bd5` - please confirm you can pull it and update the PR to pull that.  It's still a pretty big image (150 MB) - is there anything we can do that is smaller?

pmorie commented:
LGTM modulo my comment about controlling the image version ourselves

pmorie commented:
Btw, thanks for the e2e, :100:

resouer commented:
```
$ docker push gcr.io/google_containers/gitolite-http
The push refers to a repository [gcr.io/google_containers/gitolite-http] (len: 1)
604c51b84bd5: Image already exists 
8a16d946178f: Pushing [==================================================>] 3.584 kB
unknown: Access denied.
```

Anyway to solve this? @pmorie @thockin

thockin commented:
So the comment is not lost on new push: I've pushed it to gcr.io/google_containers/gitolite-http:604c51b84bd5 - please confirm you can pull it and update the PR to pull that. It's still a pretty big image (150 MB) - is there anything we can do that is smaller?

resouer commented:
The most expensive parts are `debian:wheezy` and installation of apache, based on [this](https://imagelayers.io/?images=hyzual%2Fgitolite-http:latest)

Both of them are required according to my poor knowledge about gitolite.

pmorie commented:
I briefly did some looking and found that the gitlab image is a whopping 650M.

pmorie commented:
Hrm: https://gist.github.com/MakeNowJust/262e080217aebbc628c0

pmorie commented:
Here's one that is around 75M: https://github.com/gogits/gogs/tree/master/docker

pmorie commented:
I wonder how hard it would be to make a Go program that had enough smarts to server a single repo over http (the 'dumb' protocol for git).

pmorie commented:
We could have this tiny git server and a built-in repo that you could clone with a single test file in it that you could check to determine that the git clone ran correctly.

thockin commented:
If this is all that is blocking the fix, we should roll it and come back to it later

pmorie commented:
+1 Tim.

On Tue, Jan 19, 2016 at 5:06 PM, Tim Hockin <notifications@github.com>
wrote:

> If this is all that is blocking the fix, we should roll it and come back
> to it later
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18445#issuecomment-173003118>
> .
>

pmorie commented on test/e2e/empty_dir_wrapper.go:
```diff
@@ -0,0 +1,201 @@
+var _ = Describe("EmptyDir wrapper volumes", func() {
+	f := NewFramework("secrets")
+
+	It("should becomes running [Conformance]", func() {
```
@resouer Why don't you mark this skipped with `[Skipped]` and we'll chew on how to cut the image size down in a follow-up?

cc @thockin

resouer commented:
@pmorie  Thanks for your tips, it's very helpfule! So I build a small golang web server as a fake git server (4.989 MB), thus e2e will clone a empty git repo (but with .git) with a specified repo name, it is quite enough for this test.

And please help me tag it if looks good :)

pmorie commented:
@resouer That is awesome -- put the code in test/images and we'll push to gcr.io?

resouer commented:
@pmorie What's the specific process, should I open another PR for test/images? And what will be the image name in this PR then?

thockin commented:
you can put it in this PR, but a different commit, please

On Tue, Jan 19, 2016 at 8:28 PM, Harry Zhang <notifications@github.com>
wrote:

> @pmorie <https://github.com/pmorie> What's the specific process, should I
> open another PR for test/images? And what will be the image name in this PR
> then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18445#issuecomment-173083615>
> .
>

resouer commented:
Will that image been build and push to gcr.io automatically?

thockin commented:
No, it won't auto push, and in fact, to do it sanely I was wrong.  It
should be a different PR.

Can you send me a PR that is just this image?  Additionally, make the build
drop a /GITHASH.txt file into the generated image that holds the contents
of `git rev-parse HEAD` (so we can maybe oneday rebuild it) and I will
merge that PR, build the image, and push to GCR.

Then this PR can use it.

On Tue, Jan 19, 2016 at 10:41 PM, Harry Zhang <notifications@github.com>
wrote:

> Will that image been build and push to gcr.io automatically?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18445#issuecomment-173109927>
> .
>

resouer commented:
@thockin Yep, its #19857

pmorie commented on test/e2e/empty_dir_wrapper.go:
```diff
@@ -0,0 +1,179 @@
+						},
+					},
+				},
+				Containers: []api.Container{
```
@resouer Do I understand correctly that the git repo volume setup running correctly is the thing that is under test here?

resouer replied:
Yes, otherwise, kubelet will report "volume path not found", and this pod will never becomes running.

pmorie replied:
Cool, I'm going to try it out myself now, sorry for the delay

thockin commented:
The fake git server is pushed.

resouer commented:
Updated with gcr.io image

resouer commented:
@k8s-bot e2e test this please

pmorie commented:
Thanks a lot, @resouer

jdef commented:
This PR has introduced a new binary dependency for conformance tests - that the kube nodes must have `git` installed and in `$PATH`
